appinventor/components/src/com/google/appinventor/components/runtime/CSVFile.java

```diff
@@ -0,0 +1,280 @@
+    public void SourceFile(String source) {
+        // The SourceFile property is only set from the Designer, so the
+        // only possibility is a media file. Since media file paths are
+        // distinguished by double slashes at the beginning, they need
+        // to be added.
+        ReadFile("//" + source);
+    }
+
+    @SimpleFunction(description = "Indicates source file to load data from." +
+        "Prefix the filename with / to read from a specific file on the SD card. " +
+        "for instance /myFile.txt will read the file /sdcard/myFile.txt. To read " +
+        "assets packaged with an application (also works for the Companion) start " +
+        "the filename with // (two slashes). If a filename does not start with a " +
+        "slash, it will be read from the applications private storage (for packaged " +
+        "apps) and from /sdcard/AppInventor/data for the Companion." +
+        "The results of the reading are stored in the Rows, Columns " +
+        "and ColumnNames properties of the component.")
+    public void ReadFile(String source) {
+        this.sourceFile = source;
+
+        readFromFile(sourceFile);
+    }
+
+    /**
+     * Returns a Future object which holds the CSVFile columns at the point
+     * of invoking the method.
+     *
+     * If reading is in progress, the method blocks until reading is done
+     * before returning the result.
+     *
+     * The method should be called asynchronously to prevent freezing of
+     * the main thread.
+     *
+     * The row size is contained in the method to create default values for the
+     * CharDataModel in case of an absence of columns.
+     *
+     * @param columns  List of columns to retrieve (String object entries expected)
+     * @return  Future object containing YailList of format (rowCount, columns)
+     */
+    public Future<YailList> getColumns(final YailList columns) {
+        // Submit a callable which constructs the results.
+        // The callable is only executed after all the previous
+        // tasks have been completed.
+        return threadRunner.submit(new Callable<YailList>() {
+            @Override
+            public YailList call() {
+                ArrayList<YailList> resultingColumns = new ArrayList<YailList>();
+
+                // Iterate over the specified column names
+                for (int i = 0; i < columns.size(); ++i) {
+                    // Get and add the specified column to the resulting columns list
+                    String columnName = columns.getString(i);
+                    YailList column = getColumn(columnName);
+                    resultingColumns.add(column);
+                }
+
+                // Convert result to a YailList
+                YailList csvColumns = YailList.makeList(resultingColumns);
+
+                // Final result contains the row size as the first element
+                return YailList.makeList(Arrays.asList(rows.size(), csvColumns));
```

Comment:
Return only csvColumns and establish row size in the class which uses the result instead (so as to also not import any data when no columns are specified)